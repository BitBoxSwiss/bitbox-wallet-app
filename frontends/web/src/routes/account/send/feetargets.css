.row {
    display: flex;
    flex-wrap: wrap;
}

.column {
    flex-basis: 50%;
    flex-grow: 1;
    flex-shrink: 1;
}
.column:first-child {
    flex-basis: 10%;
}

select.priority {
    border-right: none;
    padding-right: 0;
    text-transform: capitalize;
}

.fee input {
    border-left: none;
    padding-left: 0;
}

select:disabled.priority,
.fee input:disabled {
    border-color: var(--color-lightgray) !important;
}

select.priority,
.fee input {
    border-color: var(--color-mediumgray) !important;
}

.feeDescription {
    color: var(--color-secondary);
    flex: 0 0 50%;
    font-size: var(--size-small);
    font-weight: 400;
    margin-top: 0;
}
