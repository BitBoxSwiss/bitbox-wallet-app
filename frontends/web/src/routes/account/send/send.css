.container {
    margin-top: calc(var(--space-default) * 1.5);
}

.input {
    padding: var(--spacing-default);
    border-radius: 2px;
    border: solid 1px var(--color-brightgray);
    font-size: var(--size-medium);
}

.input:focus {
    outline: none;
}

.labelHalf {
    width: 49%;
}

.labelHalf:first-child {
    margin-right: 2%;
}

.inputHalf {
    width: 49%;
}

.inputFull {
    width: 100%;
}

.notAllowed {
    cursor: not-allowed;
}

.confirmationTitle {
    margin-bottom: 0;
}

.confirmationBox {
    border-radius: 2px;
    border: solid 1px var(--color-lightgray);
}

.confirmationBox .block {
    padding: var(--space-half);
    overflow: hidden;
}

.confirmationBox .block.ignorePadding {
    padding: 0;
}

.confirmationBox .block:not(:last-child) {
    border-bottom: solid 0.5px var(--color-lightgray);
}

.confirmationBox .block:last-child {
    display: flex;
    flex-direction: column;
}

.confirmationBox .block p {
    margin: 0;
}

.confirmationBox .block .half {
    padding: var(--spacing-default);
    width: 50%;
}

.confirmationBox .block .half:first-child {
    border-right: solid 0.5px var(--color-lightgray);
}

.confirmationBox .block .confirmationLabel {
    margin-bottom: var(--space-quarter);
}

.confirmationValueTable {
    border-collapse: collapse;
    font-size: var(--size-default);
    font-weight: 300;
}

.confirmationValueTable tr:first-child td {
    padding-bottom: var(--space-quarter);
}

.confirmationValueTable td {
    line-height: 1;
    vertical-align: bottom;
}

.confirmationValueTable td:first-child {
    text-align: right;
    padding-right: var(--spacing-half);
}

.confirmationValueTable tr > td:last-child {
    color: var(--color-secondary);
    font-size: var(--size-label);
    font-weight: normal;
}

.confirmationValueTable.total td {
    font-weight: 300;
    font-size: var(--size-large);
}

.confirmationValue {
    font-size: var(--size-default);
    font-weight: 300;
    line-height: 1;
    overflow: hidden;
    text-overflow: ellipsis;
    direction: rtl;
}

.maxAmount label {
    display: block !important;
    margin-right: 0 !important;
}

.maxAmount label::before {
    top: -2px !important;
}

.maxAmount label::after {
    margin-top: -3px !important;
}

.action {
    color: var(--color-blue);
    font-size: var(--size-small);
    font-weight: 300;
    line-height: 1;
    cursor: pointer;
    user-select: none;
}

.feeDescription {
    margin-top: 0;
    font-size: var(--size-small);
    font-weight: 300;
    color: var(--color-secondary);
}

.inputWithIcon {
    position: relative;
}

.inputWithIcon input {
    padding-right: calc(var(--spacing-default) * 2 + 18px);
}

.qrButton {
    border: none;
    background-color: var(--color-white);
    width: 18px;
    height: 18px;
    padding: 0;
    position: absolute;
    top: 50%;
    right: var(--spacing-default);
    cursor: pointer;
}

.qrButton img {
    width: 100%;
    height: 100%;
}

.qrVideo {
    width: 100%;
    margin-bottom: var(--spacing-default);
}