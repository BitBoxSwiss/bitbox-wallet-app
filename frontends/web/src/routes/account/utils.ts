/**
 * Copyright 2018 Shift Devices AG
 * Copyright 2021 Shift Crypto AG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { CoinCode, ScriptType, IAccount, CoinUnit } from '../../api/account';

export function findAccount(accounts: IAccount[], accountCode: string): IAccount | undefined {
  return accounts.find(({ code }) => accountCode === code);
}

export function getCryptoName(cryptoLabel: string, account?: IAccount): string {
  if (account && isBitcoinOnly(account.coinCode)) {
    return 'Bitcoin';
  }
  return cryptoLabel;
}

export function isBitcoinOnly(coinCode: CoinCode): boolean {
  switch (coinCode) {
  case 'btc':
  case 'tbtc':
    return true;
  default:
    return false;
  }
}

export const isBitcoinCoin = (coin: CoinUnit) => (coin === 'BTC') || (coin === 'TBTC') || (coin === 'sat') || (coin === 'tsat');

export function isBitcoinBased(coinCode: CoinCode): boolean {
  switch (coinCode) {
  case 'btc':
  case 'tbtc':
  case 'ltc':
  case 'tltc':
    return true;
  default:
    return false;
  }
}

export function isEthereumBased(coinCode: CoinCode): boolean {
  return coinCode === 'eth' || coinCode === 'goeth' || coinCode === 'sepeth' || coinCode.startsWith('eth-erc20-');
}

export function getCoinCode(coinCode: CoinCode): CoinCode | undefined {
  switch (coinCode) {
  case 'btc':
  case 'tbtc':
    return 'btc';
  case 'ltc':
  case 'tltc':
    return 'ltc';
  case 'eth':
  case 'goeth':
  case 'sepeth':
    return 'eth';
  }
}


export function getScriptName(scriptType: ScriptType): string {
  switch (scriptType) {
  case 'p2pkh':
    return 'Legacy (P2PKH)';
  case 'p2wpkh-p2sh':
    return 'Wrapped Segwit (P2WPKH-P2SH)';
  case 'p2wpkh':
    return 'Native segwit (bech32, P2WPKH)';
  case 'p2tr':
    return 'Taproot (bech32m, P2TR)';
  }
}

export function customFeeUnit(coinCode: CoinCode): string {
  if (isBitcoinBased(coinCode)) {
    return 'sat/vB';
  }
  if (isEthereumBased(coinCode)) {
    return 'Gwei';
  }
  return '';
}
