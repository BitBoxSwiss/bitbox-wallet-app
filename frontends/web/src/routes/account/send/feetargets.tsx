/**
 * Copyright 2018 Shift Devices AG
 * Copyright 2020 Shift Crypto AG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component, h, RenderableProps } from 'preact';
import { Input, Select } from '../../../components/forms';
import { Fiat } from '../../../components/rates/rates';
import { translate, TranslateProps } from '../../../decorators/translate';
import { apiGet } from '../../../utils/request';
import { CoinCode } from '../account';
import * as style from './feetargets.css';
import { AmountWithConversions } from './send';

export type Code = 'custom' | 'low' | 'economy' | 'normal' | 'high';

interface FeeTargetsProps {
    accountCode: CoinCode;
    disabled: boolean;
    fiatUnit: Fiat;
    proposedFee?: AmountWithConversions;
    onFeeTargetChange: (code: Code) => void;
}

export type Props = FeeTargetsProps & TranslateProps;

interface FeeTarget {
    code: Code;
}
interface State {
    feeTargets: FeeTarget[] | null;
    feeTarget?: string | null;
}

class FeeTargets extends Component<Props, State> {
    public readonly state: State = {
        feeTargets: null,
        feeTarget: null,
    };

    public componentDidMount() {
        this.updateFeeTargets(this.props.accountCode);
    }

    public componentWillReceiveProps({ accountCode }) {
        if (this.props.accountCode !== accountCode) {
            this.updateFeeTargets(accountCode);
        }
    }

    private updateFeeTargets = (accountCode: CoinCode) => {
        apiGet('account/' + accountCode + '/fee-targets')
        .then(({ feeTargets, defaultFeeTarget }: {feeTargets: FeeTarget[], defaultFeeTarget: Code}) => {
            // feeTargets.push({code: 'custom'});
            this.setState({ feeTargets });
            this.setFeeTarget(defaultFeeTarget);
        });
    }

    private handleFeeTargetChange = (event: Event) => {
        const target = event.target as HTMLSelectElement;
        const feeTargets = this.state.feeTargets;
        if (feeTargets) {
            this.setFeeTarget(feeTargets[target.selectedIndex].code);
        }
    }

    private setFeeTarget = (feeTarget: Code) => {
        this.setState({ feeTarget });
        this.props.onFeeTargetChange(feeTarget);
    }

    public render(
    {
        t,
        disabled,
        fiatUnit,
        proposedFee,
    }: RenderableProps<Props>,
    {
        feeTargets,
        feeTarget,
    }: State) {
        if (feeTargets === null) {
            return (
                <Input
                    label={t('send.priority')}
                    id="feetarget"
                    placeholder={t('send.feeTarget.placeholder')}
                    disabled
                    transparent />
            );
        }
        if (feeTargets.length === 0) {
            return null;
        }

        return (
            <div className={style.row}>
                <div className={style.column}>
                    <Select
                        className={style.priority}
                        label={t('send.priority')}
                        id="feeTarget"
                        disabled={disabled}
                        onChange={this.handleFeeTargetChange}
                        selected={feeTarget}
                        options={feeTargets.map(({ code }) => {
                            return {
                                value: code,
                                text: t(`send.feeTarget.label.${code}`),
                            };
                        })} />
                </div>
                <div className={style.column}>
                    <Input
                        align="right"
                        className={style.fee}
                        disabled={feeTarget !== 'custom'}
                        label={t('send.fee.label')}
                        id="proposedFee"
                        placeholder={feeTarget === 'custom' ? t('send.fee.customPlaceholder') : t('send.fee.placeholder')}
                        transparent
                        value={proposedFee && proposedFee.amount + ' ' + proposedFee.unit + (proposedFee.conversions ? ' = ' + proposedFee.conversions[fiatUnit] + ' ' + fiatUnit : '')}
                        />
                </div>
                { feeTarget && (
                    <div>
                        <label>{t('send.feeTarget.estimate')}</label>
                        <p class={style.feeDescription}>{t('send.feeTarget.description.' + feeTarget)}</p>
                    </div>
                )}
            </div>
        );
    }
}

const TranslatedFeeTargets = translate<FeeTargetsProps>()(FeeTargets);
export { TranslatedFeeTargets as FeeTargets };
