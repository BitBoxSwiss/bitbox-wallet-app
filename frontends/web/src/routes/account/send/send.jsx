/**
 * Copyright 2018 Shift Devices AG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component, h } from 'preact';
import { route } from 'preact-router';
import { translate } from 'react-i18next';
import { apiGet, apiPost } from '../../../utils/request';
import { apiWebsocket } from '../../../utils/websocket';
import { debug } from '../../../utils/env';
import { Button, ButtonLink, Checkbox, Input } from '../../../components/forms';
import { Guide } from '../../../components/guide/guide';
import { Entry } from '../../../components/guide/entry';
import { store as fiat } from '../../../components/rates/rates';
import { alertUser } from '../../../components/alert/Alert';
import Header from '../../../components/header/Header';
import Status from '../../../components/status/status';
import WaitDialog from '../../../components/wait-dialog/wait-dialog';
import Balance from '../../../components/balance/balance';
import FeeTargets from './feetargets';
import UTXOs from './utxos';
import approve from '../../../assets/icons/checked.svg';
import reject from '../../../assets/icons/cancel.svg';
import * as style from './send.css';

@translate()
export default class Send extends Component {

    constructor(props) {
        super(props);

        this.state = {
            balance: null,
            amount: null,
            feeTarget: null,
            proposedFee: null,
            proposedAmount: null,
            proposedTotal: null,
            valid: false,
            addressError: null,
            amountError: null,
            sendAll: false,
            isConfirming: false,
            isSent: false,
            isAborted: false,
            paired: null,
            lock: null,
            fiatAmount: null,
            fiatUnit: fiat.state.active,
            signProgress: null,
            signConfirm: null, // show visual BitBox in dialog when instructed to sign.
            coinControl: false,
            activeCoinControl: false,
        };
        this.selectedUTXOs = [];
    }

    coinSupportsCoinControl = () => {
        const account = this.getAccount();
        if (!account) {
            return false;
        }
        switch (account.coinCode) {
        case 'btc':
        case 'tbtc':
        case 'ltc':
        case 'tltc':
            return true;
        default:
            return false;
        }
    }

    componentDidMount() {
        apiGet(`account/${this.props.code}/balance`).then(balance => this.setState({ balance }));
        if (this.props.deviceIDs.length > 0) {
            apiGet('devices/' + this.props.deviceIDs[0] + '/paired').then((paired) => {
                apiGet('devices/' + this.props.deviceIDs[0] + '/info').then(({ lock }) => {
                    this.setState({ paired, lock });
                });
            });
        }
        if (this.coinSupportsCoinControl()) {
            apiGet('config').then(config => this.setState({ coinControl: !!(config.frontend || {}).coinControl }));
        }
        this.unsubscribe = apiWebsocket(({ type, data, meta }) => {
            switch (type) {
            case 'device':
                switch (data) {
                case 'signProgress':
                    this.setState({ signProgress: meta, signConfirm: null });
                    break;
                case 'signConfirm':
                    this.setState({ signConfirm: true });
                    break;
                }
                break;
            }
        });
    }

    componentWillMount() {
        this.registerEvents();
    }

    componentWillUnmount() {
        this.unregisterEvents();
        if (this.unsubscribe) {
            this.unsubscribe();
        }
    }

    registerEvents = () => {
        document.addEventListener('keydown', this.handleKeyDown);
    }

    unregisterEvents = () => {
        document.removeEventListener('keydown', this.handleKeyDown);
    }

    handleKeyDown = e => {
        if (e.keyCode === 27) {
            console.log('send.jsx route to /'); // eslint-disable-line no-console
            route(`/account/${this.props.code}`);
        }
    }

    send = () => {
        this.setState({ signProgress: null, isConfirming: true });
        apiPost('account/' + this.getAccount().code + '/sendtx', this.txInput()).then(result => {
            if (result.success) {
                this.setState({
                    sendAll: false,
                    isConfirming: false,
                    isSent: true,
                    recipientAddress: null,
                    proposedAmount: null,
                    proposedFee: null,
                    proposedTotal: null,
                    fiatAmount: null,
                    amount: null,
                });
                if (this.utxos) {
                    this.utxos.getWrappedInstance().clear();
                }
                setTimeout(() => this.setState({ isSent: false, isConfirming: false }), 5000);
            } else {
                this.setState({
                    isAborted: true,
                });
                setTimeout(() => this.setState({ isAborted: false }), 5000);
            }
            // The following method allows pressing escape again.
            this.setState({ isConfirming: false,signProgress: null, signConfirm: null });
        }).catch(() => {
            this.setState({ isConfirming: false,signProgress: null, signConfirm: null });
        });
    }

    txInput = () => ({
        address: this.state.recipientAddress,
        amount: this.state.amount,
        feeTarget: this.state.feeTarget || '',
        sendAll: this.state.sendAll ? 'yes' : 'no',
        selectedUTXOs: Object.keys(this.selectedUTXOs),
    })

    sendDisabled = () => {
        const txInput = this.txInput();
        return !txInput.address || this.state.feeTarget === null || (txInput.sendAll === 'no' && !txInput.amount);
    }

    validateAndDisplayFee = updateFiat => {
        this.setState({
            proposedTotal: null,
            addressError: null,
            amountError: null,
        });
        if (this.sendDisabled()) {
            return;
        }
        const txInput = this.txInput();
        apiPost('account/' + this.getAccount().code + '/tx-proposal', txInput).then(result => {
            this.setState({ valid: result.success });
            if (result.success) {
                this.setState({
                    addressError: null,
                    amountError: null,
                    proposedFee: result.fee,
                    proposedAmount: result.amount,
                    proposedTotal: result.total,
                });
                if (updateFiat) {
                    this.convertToFiat(result.amount.amount);
                }
            } else {
                const errorCode = result.errorCode;
                switch (errorCode) {
                case 'invalidAddress':
                    this.setState({ addressError: this.props.t('send.error.invalidAddress') });
                    break;
                case 'invalidAmount':
                case 'insufficientFunds':
                    this.setState({ amountError: this.props.t(`send.error.${errorCode}`) });
                    break;
                default:
                    this.setState({ proposedFee: null });
                    if (errorCode) {
                        this.unregisterEvents();
                        alertUser(errorCode, this.registerEvents);
                    }
                }
            }
        }).catch(() => {
            this.setState({ valid: false });
        });
    }

    handleFormChange = event => {
        let value = event.target.value;
        if (event.target.type === 'checkbox') {
            value = event.target.checked;
        }
        if (event.target.id === 'sendAll') {
            if (!value) {
                this.convertToFiat(this.state.amount);
            }
        } else if (event.target.id === 'amount') {
            this.convertToFiat(value);
        }
        this.setState({ [event.target.id]: value });
        this.validateAndDisplayFee(true);
    }

    handleFiatInput = event => {
        const value = event.target.value;
        this.setState({ fiatAmount: value });
        this.convertFromFiat(value);
    }

    convertToFiat = value => {
        if (value) {
            let coinUnit = this.getAccount().coinCode.toUpperCase();
            if (coinUnit.length === 4 && coinUnit.startsWith('T')) {
                coinUnit = coinUnit.substring(1);
            }
            apiGet(`coins/convertToFiat?from=${coinUnit}&to=${this.state.fiatUnit}&amount=${value}`)
                .then(data => {
                    if (data.success) {
                        this.setState({ fiatAmount: data.fiatAmount });
                    } else {
                        this.setState({ amountError: this.props.t(`send.error.invalidAmount`) });
                    }
                });
        } else {
            this.setState({ fiatAmount: null });
        }
    }

    convertFromFiat = value => {
        if (value) {
            let coinUnit = this.getAccount().coinCode.toUpperCase();
            if (coinUnit.length === 4 && coinUnit.startsWith('T')) {
                coinUnit = coinUnit.substring(1);
            }
            apiGet(`coins/convertFromFiat?from=${this.state.fiatUnit}&to=${coinUnit}&amount=${value}`)
                .then(data => {
                    if (data.success) {
                        this.setState({ amount: data.amount });
                        this.validateAndDisplayFee(false);
                    } else {
                        this.setState({ amountError: this.props.t(`send.error.invalidAmount`) });
                    }
                });
        } else {
            this.setState({ amount: null });
        }
    }

    sendAll = event => {
        this.handleFormChange(event);
    }

    sendToSelf = event => {
        apiGet('account/' + this.getAccount().code + '/receive-addresses').then(receiveAddresses => {
            this.setState({ recipientAddress: receiveAddresses[0].address });
            this.handleFormChange(event);
        });
    }

    feeTargetChange = feeTarget => {
        this.setState({ feeTarget });
        this.validateAndDisplayFee(this.state.sendAll);
    }

    onSelectedUTXOsChange = selectedUTXOs => {
        this.selectedUTXOs = selectedUTXOs;
        this.validateAndDisplayFee(true);
    }

    getAccount() {
        if (!this.props.accounts) return null;
        return this.props.accounts.find(({ code }) => code === this.props.code);
    }

    toggleCoinControl = () => {
        this.setState(({ activeCoinControl }) => {
            if (activeCoinControl && this.utxos) {
                this.utxos.getWrappedInstance().clear();
            }
            return { activeCoinControl: !activeCoinControl };
        });
    }

    render({
        t,
        code,
    }, {
        balance,
        proposedFee,
        proposedTotal,
        recipientAddress,
        proposedAmount,
        valid,
        amount,
        fiatAmount,
        fiatUnit,
        sendAll,
        feeTarget,
        isConfirming,
        isSent,
        isAborted,
        addressError,
        amountError,
        paired,
        lock,
        signProgress,
        signConfirm,
        coinControl,
        activeCoinControl,
    }) {
        const account = this.getAccount();
        if (!account) return null;

        const confirmPrequel = signProgress && signProgress.steps > 1 ? (
            <span>
                {t('send.signprogress.description', {
                    steps: signProgress.steps,
                })}<br />
                {t('send.signprogress.label')}: {signProgress.step}/{signProgress.steps}
            </span>
        ) : null;
        return (
            <div class="contentWithGuide">
                <div class="container">
                    <Status type="warning">
                        {paired === false && t('warning.sendPairing')}
                    </Status>
                    <Header title={<h2>{t('send.title')}</h2>} {...this.props}>
                        <Balance
                            t={t}
                            balance={balance} />
                        {
                            coinControl ? (
                                <div style="align-self: flex-end;">
                                    <Button onClick={this.toggleCoinControl} primary>{t('send.toggleCoinControl')}</Button>
                                </div>
                            ) : null
                        }
                    </Header>
                    <div class="innerContainer scrollableContainer">
                        <div class="content padded">
                            {
                                coinControl && (
                                    <UTXOs
                                        active={activeCoinControl}
                                        ref={ref => this.utxos = ref}
                                        accountCode={account.code}
                                        onChange={this.onSelectedUTXOsChange} />
                                )
                            }
                            <div class="row">
                                <Input
                                    label={t('send.address.label')}
                                    placeholder={t('send.address.placeholder')}
                                    id="recipientAddress"
                                    error={addressError}
                                    onInput={this.handleFormChange}
                                    value={recipientAddress}
                                    autofocus
                                />
                                { debug && (
                                    <span id="sendToSelf" className={style.action} onClick={this.sendToSelf}>
                                        Send to self
                                    </span>
                                ) }
                            </div>
                            <div class="row">
                                <div class="flex flex-1 flex-row flex-between flex-items-center spaced">
                                    <Input
                                        label={t('send.amount.label')}
                                        id="amount"
                                        onInput={this.handleFormChange}
                                        disabled={sendAll}
                                        error={amountError}
                                        value={sendAll ? proposedAmount && proposedAmount.amount : amount}
                                        placeholder={`${t('send.amount.placeholder')} ` + (balance && `(${balance.available.unit})`)} />
                                    <Input
                                        label={fiatUnit}
                                        id="fiatAmount"
                                        onInput={this.handleFiatInput}
                                        disabled={sendAll}
                                        error={amountError}
                                        value={fiatAmount}
                                        placeholder={`${t('send.amount.placeholder')} (${fiatUnit})`} />
                                </div>
                                <div class="flex flex-1 flex-row flex-between flex-items-center spaced">
                                    <Checkbox
                                        label={t('send.maximum')}
                                        id="sendAll"
                                        onChange={this.sendAll}
                                        checked={sendAll}
                                        className={style.maxAmount} />
                                </div>
                            </div>
                            <div class="row">
                                <div class="flex flex-1 flex-row flex-between flex-items-center spaced">
                                    <FeeTargets
                                        label={t('send.feeTarget.label')}
                                        placeholder={t('send.feeTarget.placeholder')}
                                        accountCode={account.code}
                                        disabled={!amount && !sendAll}
                                        onFeeTargetChange={this.feeTargetChange} />
                                    <Input
                                        label={t('send.fee.label')}
                                        value={proposedFee ? proposedFee.amount + ' ' + proposedFee.unit + (proposedFee.conversions ? ' = ' + proposedFee.conversions[fiatUnit] + ' ' + fiatUnit : '') : null}
                                        placeholder={feeTarget === 'custom' ? t('send.fee.customPlaceholder') : t('send.fee.placeholder')}
                                        disabled={feeTarget !==  'custom'}
                                        transparent />
                                    {/*
                                    <Input
                                        label={t('send.customFee.label')}
                                        placeholder={t('send.customFee.placeholder')}
                                        disabled
                                    />
                                    */}
                                </div>
                                <p class={style.feeDescription}>{feeTarget && t('send.feeTarget.description.' + feeTarget) || ''}</p>
                            </div>
                            <div class="row buttons flex flex-row flex-between flex-start">
                                <ButtonLink
                                    secondary
                                    href={`/account/${code}`}>
                                    {t('button.back')}
                                </ButtonLink>
                                <Button primary onClick={this.send} disabled={this.sendDisabled() || !valid}>
                                    {t('send.button')}
                                </Button>
                            </div>
                        </div>
                    </div>
                    {
                        isConfirming && (
                            <WaitDialog
                                title={t('send.confirm.title')}
                                prequel={confirmPrequel}
                                paired={paired}
                                lock={lock}
                                touchConfirm={signConfirm}
                                includeDefault>
                                <div class={style.confirmationBox}>
                                    <div class={style.block}>
                                        <p class={['label', style.confirmationLabel, 'first'].join(' ')}>
                                            {t('send.address.label')}
                                        </p>
                                        <p class={style.confirmationValue}>{recipientAddress || 'N/A'}</p>
                                    </div>
                                    <div class={['flex flex-row flex-start', style.block, style.ignorePadding].join(' ')}>
                                        <div class={style.half}>
                                            <p class={['label', style.confirmationLabel].join(' ')}>
                                                {t('send.amount.label')}
                                            </p>
                                            <table class={style.confirmationValueTable} align="right">
                                                <tr>
                                                    <td>{proposedAmount && proposedAmount.amount || 'N/A'}</td>
                                                    <td>{proposedAmount && proposedAmount.unit || 'N/A'}</td>
                                                </tr>
                                                {
                                                    proposedAmount && proposedAmount.conversions && (
                                                        <tr>
                                                            <td>{proposedAmount.conversions[fiatUnit]}</td>
                                                            <td>{fiatUnit}</td>
                                                        </tr>
                                                    )
                                                }
                                            </table>
                                        </div>
                                        <div class={style.half}>
                                            <p class={['label', style.confirmationLabel].join(' ')}>
                                                {t('send.fee.label')}
                                                {' '}
                                                ({feeTarget})
                                            </p>
                                            <table class={style.confirmationValueTable} align="right">
                                                <tr>
                                                    <td>{proposedFee && proposedFee.amount || 'N/A'}</td>
                                                    <td>{proposedFee && proposedFee.unit || 'N/A'}</td>
                                                </tr>
                                                {
                                                    proposedFee && proposedFee.conversions && (
                                                        <tr>
                                                            <td>{proposedFee.conversions[fiatUnit]}</td>
                                                            <td>{fiatUnit}</td>
                                                        </tr>
                                                    )
                                                }
                                            </table>
                                        </div>
                                    </div>
                                    {
                                        this.selectedUTXOs.length !== 0 && (
                                            <div class={style.block}>
                                                <p class={['label', style.confirmationLabel].join(' ')}>
                                                    {t('send.confirm.selected-coins')}
                                                </p>
                                                {
                                                    Object.keys(this.selectedUTXOs).map((uxto, i) => (
                                                        <p class={style.confirmationValue} key={`selectedCoin-${i}`}>{uxto}</p>
                                                    ))
                                                }
                                            </div>
                                        )
                                    }
                                    <div class={style.block}>
                                        <p class={['label', style.confirmationLabel].join(' ')}>
                                            {t('send.confirm.total')}
                                        </p>
                                        <div>
                                            <table class={[style.confirmationValueTable, style.total].join(' ')} align="right">
                                                <tr>
                                                    <td>{proposedTotal && proposedTotal.amount || 'N/A'}</td>
                                                    <td>{proposedTotal && proposedTotal.unit || 'N/A'}</td>
                                                </tr>
                                                {
                                                    proposedTotal && proposedTotal.conversions && (
                                                        <tr>
                                                            <td>{proposedTotal.conversions[fiatUnit]}</td>
                                                            <td>{fiatUnit}</td>
                                                        </tr>
                                                    )
                                                }
                                            </table>
                                        </div>
                                    </div>
                                </div>
                            </WaitDialog>
                        )
                    }
                    {
                        isSent && (
                            <WaitDialog>
                                <div class="flex flex-row flex-center flex-items-center text-bold">
                                    <img src={approve} alt="Success" style="height: 40px; margin-right: 1rem;" />{t('send.success')}
                                </div>
                            </WaitDialog>
                        )
                    }
                    {
                        isAborted && (
                            <WaitDialog>
                                <div class="flex flex-row flex-center flex-items-center text-bold">
                                    <img src={reject} alt="Abort" style="height: 40px; margin-right: 1rem;" />{t('send.abort')}
                                </div>
                            </WaitDialog>
                        )
                    }
                </div>
                <Guide>
                    <Entry key="guide.send.whyFee" entry={t('guide.send.whyFee')} />
                    <Entry key="guide.send.priority" entry={t('guide.send.priority')} />
                    <Entry key="guide.send.fee" entry={t('guide.send.fee')} />
                    <Entry key="guide.send.revert" entry={t('guide.send.revert')} />
                </Guide>
            </div>
        );
    }
}
