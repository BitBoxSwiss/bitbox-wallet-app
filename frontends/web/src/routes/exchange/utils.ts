/**
 * Copyright 2022-2024 Shift Crypto AG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { IAccount } from '@/api/account';
import { getExchangeSupported, TExchangeName } from '@/api/exchanges';

/**
 * Gets formatted name for exchange.
 */
export const getExchangeFormattedName = (
  name: TExchangeName,
) => {
  switch (name) {
  case 'moonpay':
    return 'MoonPay';
  case 'pocket':
    return 'Pocket';
  case 'btcdirect':
    return 'BTC Direct';
  case 'btcdirect-otc':
    return 'BTC Direct\'s Private Trading Desk';
  case 'bitrefill':
    return 'Bitrefill';
  }
};

/**
 * Filters a given accounts list, keeping only the accounts supported by at least one exchange.
 */
export const getExchangeSupportedAccounts = async (accounts: IAccount[]): Promise<IAccount[]> => {
  const accountsWithFalsyValue = await Promise.all(
    accounts.map(async (account) => {
      const supported = await getExchangeSupported(account.code)();
      return supported.exchanges.length ? account : false;
    })
  );
  return accountsWithFalsyValue.filter(result => result) as IAccount[];
};
