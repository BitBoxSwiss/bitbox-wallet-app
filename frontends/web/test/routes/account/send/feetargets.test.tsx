/**
 * Copyright 2020 Shift Crypto AG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import 'jest';
import '../../../matchmediastub';

import { h } from 'preact';
import { deep } from 'preact-render-spy';

jest.mock('../../../../src/utils/request');
import { apiGet } from '../../../../src/utils/request';

jest.mock('../../../../src/i18n/i18n');

jest.mock('../../../../src/decorators/translate', () => ({
    // this mock makes sure any components using the translate HoC receive the t function as a prop
    translate: () => Component => {
        Component.defaultProps = { ...Component.defaultProps, t: () => '' };
        return Component;
    },
}));

import { FeeTargets, Props } from '../../../../src/routes/account/send/feetargets';

describe('routes/account/send/feetargets', () => {
    it('should match the snapshot', () => {
        (apiGet as jest.Mock).mockResolvedValue({
            defaultFeeTarget: 'economy',
            feeTargets: [{ code: 'economy' }],
        });

        const fee = deep<Props, {}>(
            <FeeTargets
                accountCode="btc"
                disabled={false}
                fiatUnit="USD"
                onFeeTargetChange={jest.fn()} />,
        );
        expect(fee).toMatchSnapshot();
    });

    it('should call onFeeTargetChange', () => {
        const apiGetMock = (apiGet as jest.Mock).mockResolvedValue({
            defaultFeeTarget: 'normal',
            feeTargets: [
                { code: 'low' },
                { code: 'economy' },
                { code: 'normal' },
                { code: 'high' },
            ],
        });

        const onFeeTargetChangeCB = jest.fn();

        deep<Props, {}>(
            <FeeTargets
                accountCode="btc"
                disabled={false}
                fiatUnit="USD"
                onFeeTargetChange={onFeeTargetChangeCB} />,
        );
        expect(apiGetMock).toHaveBeenCalled();
        expect(onFeeTargetChangeCB).toHaveBeenCalledWith('normal');
    });
});
