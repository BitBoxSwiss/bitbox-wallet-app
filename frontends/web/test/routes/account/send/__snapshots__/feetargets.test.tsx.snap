// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes/account/send/feetargets should match the snapshot 1`] = `
preact-render-spy (1 nodes)
-------
<div
  class="input align-left  isTransparent"
  style=""
>
  <input
    autocomplete="off"
    autocorrect="off"
    spellcheck={false}
    type="text"
    id="feetarget"
    placeholder=""
    disabled={true}
   />
</div>

`;
